lib/web/apiserver_test.go

```diff
@@ -1476,63 +1476,45 @@ func TestActiveSessions(t *testing.T) {
 	require.Equal(t, s.server.ClusterName(), sess.ClusterName)
 }
 
-// DELETE IN: 5.0.0
-// Tests the code snippet from apiserver.(*Handler).siteSessionGet/siteSessionsGet
-// that tests empty ClusterName and ServerHostname gets set.
-func TestEmptySessionClusterHostnameIsSet(t *testing.T) {
+func TestActiveTrackers(t *testing.T) {
 	t.Parallel()
 	s := newWebSuite(t)
-	nodeClient, err := s.server.NewClient(auth.TestBuiltin(types.RoleNode))
-	require.NoError(t, err)
-
-	// Create a session with empty ClusterName.
-	sess1 := session.Session{
-		ClusterName:    "",
-		ServerID:       string(session.NewID()),
-		ID:             session.NewID(),
-		Namespace:      apidefaults.Namespace,
-		Login:          "foo",
-		Created:        time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC),
-		LastActive:     time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC),
-		TerminalParams: session.TerminalParams{W: 100, H: 100},
-	}
-	err = nodeClient.CreateSession(sess1)
-	require.NoError(t, err)
-
-	// Retrieve the session with the empty ClusterName.
-	pack := s.authPack(t, "baz")
-	res, err := pack.clt.Get(s.ctx, pack.clt.Endpoint("webapi", "sites", s.server.ClusterName(), "sessions", sess1.ID.String()), url.Values{})
-	require.NoError(t, err)
+	sid := session.NewID()
+	pack := s.authPack(t, "foo")
 
-	// Test that empty ClusterName and ServerHostname got set.
-	var sessionResult *session.Session
-	err = json.Unmarshal(res.Bytes(), &sessionResult)
+	ws, err := s.makeTerminal(t, pack, sid)
 	require.NoError(t, err)
-	require.Equal(t, s.server.ClusterName(), sessionResult.ClusterName)
-	require.Equal(t, sess1.ServerID, sessionResult.ServerHostname)
+	t.Cleanup(func() { require.NoError(t, ws.Close()) })
```

Comment:
Can the `t.Cleanup` call go inside `makeTerminal` instead?

Reply:
I don't think that's a good idea. `makeTerminal` may be a test function but that doesn't exclude that the caller may want to manually call `Close()` for testing reasons. I believe it's more beneficial to give the caller control over it for that reason.

Reply:
Hmmm. It's not always easy to tell if you need to remember to clean something up yourself or if the test will do it for you. Additionally, failure to properly clean up has caused a number of flaky tests.

The general convention I've seen is that if some constructor func takes a `*testing.T`, then it will:

- not return an error (as it can just fail the test)
- ensure that cleanup actions are scheduled

If you want to make it the caller's responsibility to call close, then I would change the signature of `makeTerminal` such that it returns `(Terminal, error)`, and then the caller has to assert on the return values *and* handle cleanup.

Reply:
If that's convention, then sure, that's fine. I've changed it around so that `makeTerminal` does not take a `testing.T` and acts like a regular fn so the caller will have to handle cleanup & errors.